### PR TITLE
Vault-16367: add support RSA padding scheme pkcs1v15 for encryption

### PR DESCRIPTION
**Pull request by marcellanz**
This PR implements the feature requested by issue: https://github.com/hashicorp/vault/issues/16367 with support for the transit secrets engine to choose a RSA padding scheme by the `padding_scheme` parameter for `encryption`, `decryption`, `datakey` and `rewrap`.

This PR includes UI support to choose the padding scheme for RSA keys.

### CODE REVIEW

**Comment by cipherboy** [comment]
Hey there @marcellanz  -- terribly sorry about the delay. Originally I think there was some discussion over the scope of this and if it fit well with our goals for Transit. The good news is, I'd like to see this land in the next (1.15) release. :-) 

If you're still up for it, I'd be happy to work with you to get this reviewed and merged!

**Comment by cipherboy** [comment]
@marcellanz Cool! I think a rebase would be the first step; we've added a few features since then that might be causing these conflicts: managed keys and public-key only asymmetric crypto. 

I'll add some commentary in the PR :-)

File: sdk/helper/keysutil/policy.go
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:821]: I should double check I haven't violated this myself recently, but in general, we wish to avoid modifying these functions as they lie in our public `sdk/` space. 

I think you could change this to `Encrypt->EncryptWithAlgorithm` and have the former call the latter with `oaep` by default for consistency perhaps...? 

Then if/when we introduce ECEIS based schemes, I think it might make the most sense?

File: sdk/helper/keysutil/policy.go
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:927]: Same as above :-)

File: builtin/logical/transit/path_datakey.go
Comment (cipherboy) [inline builtin/logical/transit/path_datakey.go:35]: It'd be great to include `Default: "oaep"` here for consistency if we could :-) When you call `d.Get(...)` below, it'll populate it with `oaep`.

Comment (cipherboy) [reply builtin/logical/transit/path_datakey.go:35]: This applies in a few places as well below :-)

Comment (cipherboy) [reply builtin/logical/transit/path_datakey.go:35]: If I were to nitpick, for consistency with [sign/verify I might suggest calling this `encryption_algorithm`](https://github.com/hashicorp/vault/blob/main/builtin/logical/transit/path_sign_verify.go#L150-L154), as it'll likely be more widely used than just RSA here. `padding_scheme` is definitely more precise, but boxes us into a corner w.r.t. EC algorithms.

**Review by cipherboy — COMMENTED** [review]
I should get @hellobontempo to review the Transit UI additions but otherwise this is looking good to me with the one API change noted. 

Rest are tbh mostly nits.

**Comment by marcellanz** [comment]
@cipherboy thanks for the review and comments; I'll have a look this evening on them and make it mergeable.

**Comment by hellobontempo** [comment]
> I should get @hellobontempo to review the Transit UI additions but otherwise this is looking good to me with the one API change noted.
> 
> Rest are tbh mostly nits.

👋 I added the `ui` label which should run the ui tests when you push to this branch next (otherwise they're skipped) . First pass, things look good. But I won't have time to give it a more thorough review until tomorrow!

File: sdk/helper/keysutil/policy.go
Comment (marcellanz) [reply sdk/helper/keysutil/policy.go:927]: Same as with `Encrypt`

File: builtin/logical/transit/path_datakey.go
Comment (marcellanz) [reply builtin/logical/transit/path_datakey.go:35]: > I might suggest calling this `encryption_algorithm`

I think too, this makes a lot of sense and is consistent with the existing API.

File: sdk/helper/keysutil/policy.go
Comment (cipherboy) [reply sdk/helper/keysutil/policy.go:821]: No, I don't think our SDK has much documentation sadly. I think it'd be cool to get better inline godoc in the future, but we don't have [much presently](https://pkg.go.dev/github.com/hashicorp/vault/sdk/helper/keysutil). 

(If you want to work on this, happy to review such PRs in the future too!)

**Comment by marcellanz** [comment]
@cipherboy I've corrected most comments from you above and made it buildable with an up to date branch. I'll push that soon. Open is the naming change from `padding_scheme` to `encryption_algorithm` which I think is reasonable and consistent with other parameters.

**Comment by cipherboy** [comment]
@marcellanz Cool, ty! I think this might not make 1.14 sadly at this point, but 1.15 branch is open so I can merge it there when you get the updates pushed. :-)

File: ui/app/templates/components/transit-key-action/decrypt.hbs
Comment (hellobontempo) [inline ui/app/templates/components/transit-key-action/decrypt.hbs:33]: same comment here as above ☝️

File: ui/app/templates/components/transit-key-action/decrypt.hbs
Comment (hellobontempo) [inline ui/app/templates/components/transit-key-action/decrypt.hbs:45]: and here! 😄

File: ui/app/templates/components/transit-key-action/encrypt.hbs
Comment (hellobontempo) [inline ui/app/templates/components/transit-key-action/encrypt.hbs:50]: also this file and line 62

File: ui/app/templates/components/transit-key-action/rewrap.hbs
Comment (hellobontempo) [inline ui/app/templates/components/transit-key-action/rewrap.hbs:37]: last one!

**Review by hellobontempo — COMMENTED** [review]
Thanks for tackling the UI part of this work as well! 🥇 The formatting looks great - thanks for following our current linting patterns! 🤩 

I left some comments! Nothing blocking, though. It would be great to see a test added in `transit-key-actions-test.js` confirming that `padding_scheme` only renders for the desired key type - if that's something you have time to include

**Comment by cipherboy** [comment]
\o hey @marcellanz -- wanted to let you know 1.14 just GA'd, so we're happy to merge this to 1.15 branch when you get the updates in btw :-) No rush!

File: website/content/api-docs/secret/transit.mdx
Comment (schavis) [inline website/content/api-docs/secret/transit.mdx:513]:
```suggestion
- `padding_scheme` `(string: "oaep")` – Specifies the RSA encryption padding
  scheme for RSA keys. Must be one of the following supported signature types:
```

Style correction: write in active voice

File: website/content/api-docs/secret/transit.mdx
Comment (schavis) [inline website/content/api-docs/secret/transit.mdx:624]:
```suggestion
- `padding_scheme` `(string: "oaep")` – Specifies the RSA decryption padding
  scheme for RSA keys. Must be one of the following supported signature types:
```

File: website/content/api-docs/secret/transit.mdx
Comment (schavis) [inline website/content/api-docs/secret/transit.mdx:701]:
```suggestion
- `padding_scheme` `(string: "oaep")` – Specifies the RSA re-encryption padding
  scheme for RSA keys. Must be one of the following supported signature types:
```

**Review by schavis — CHANGES_REQUESTED** [review]

**Review by schavis — COMMENTED** [review]
Docs LGTM, feel free to merge once ENG signs off.

**Comment by hashishaw** [comment]
No rush, but once the merge conflicts are resolved I'm happy to give it a final look-over from the UI perspective!

**Comment by dj-thd2** [comment]
Hey everyone, is this PR still alive? This feature would be very useful to decrypt data which was encrypted with plain openssl rsautl CLI, which by default use pkcs1v15 so it's not compatible with the current transit engine. If @marcellanz is busy atm, I would be pleased to update it and shave off the pending issues to let it roll forward :)

**Comment by marcellanz** [comment]
> Hey everyone, is this PR still alive? This feature would be very useful to decrypt data which was encrypted with plain openssl rsautl CLI, which by default use pkcs1v15 so it's not compatible with the current transit engine. If @marcellanz is busy atm, I would be pleased to update it and shave off the pending issues to let it roll forward :)

Hi @dj-thd2 – alive, planned to continue soon. Let us solve merge conflicts and then seen how it fits.

**Comment by hsimon-hashicorp** [comment]
> > Hey everyone, is this PR still alive? This feature would be very useful to decrypt data which was encrypted with plain openssl rsautl CLI, which by default use pkcs1v15 so it's not compatible with the current transit engine. If @marcellanz is busy atm, I would be pleased to update it and shave off the pending issues to let it roll forward :)
> 
> Hi @dj-thd2 – alive, planned to continue soon. Let us solve merge conflicts and then seen how it fits.

Hi there, just wanting to check on this and see how things are going. Thanks!

**Pull request closed without merging**
